### PR TITLE
feat: Panel bed specific conditions

### PR DESCRIPTION
**Pull request by ivadym**
### This PR:

Allows metric filtering using capture kit specific QC conditions.
- It considers the default conditions if there is no panel bed specific requirements.
- It overrides the default (and common) requirements if there are filtering conditions for that capture kit.

#### Changed:
- The validation is done across all the samples now and not only until the first ValueError rise.

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

File: BALSAMIC/constants/quality_check_reporting.py [hunk 1]
```diff
@@ -50,70 +50,106 @@
...
-                "PCT_TARGET_BASES_1000X": {"condition": None},
-                "FOLD_80_BASE_PENALTY": {"condition": {"norm": "lt", "threshold": 5.0}},
-                "PCT_OFF_BAIT": {"condition": {"norm": "lt", "threshold": 0.45}},
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEDIAN_TARGET_COVERAGE": {"condition": None},
+                    "PCT_TARGET_BASES_50X": {"condition": None},
+                    "PCT_TARGET_BASES_100X": {"condition": None},
+                    "PCT_TARGET_BASES_250X": {"condition": None},
+                    "PCT_TARGET_BASES_500X": {"condition": None},
+                    "PCT_TARGET_BASES_1000X": {"condition": None},
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:75]:
```suggestion

```

According to the amdoc 2046, production doesn't consider `PCT_Target_Bases` for QC check

File: BALSAMIC/constants/quality_check_reporting.py [hunk 2]
```diff
@@ -50,70 +50,106 @@
...
+                    "MEAN_TARGET_COVERAGE": {"condition": None},
+                    "FOLD_80_BASE_PENALTY": {"condition": None},
+                    "PCT_OFF_BAIT": {"condition": None},
+                },
+            },
+            "gicfdna_3.1_hg19_design.bed": {
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEAN_TARGET_COVERAGE": {
+                        "condition": {"norm": "gt", "threshold": 500.0}
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:84]:
```suggestion
                        "condition": {"norm": "gt", "threshold": 1000.0}
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 3]
```diff
@@ -50,70 +50,106 @@
...
+                    "PCT_TARGET_BASES_500X": {"condition": None},
+                    "PCT_TARGET_BASES_1000X": {"condition": None},
+                    "MEAN_TARGET_COVERAGE": {"condition": None},
+                    "FOLD_80_BASE_PENALTY": {"condition": None},
+                    "PCT_OFF_BAIT": {"condition": None},
+                },
+            },
+            "gicfdna_3.1_hg19_design.bed": {
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEAN_TARGET_COVERAGE": {
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:83]:
```suggestion
                    "MEDIAN_TARGET_COVERAGE": {
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 4]
```diff
@@ -50,70 +50,106 @@
...
+                        "condition": {"norm": "gt", "threshold": 500.0}
+                    },
+                    "FOLD_80_BASE_PENALTY": {
+                        "condition": {"norm": "lt", "threshold": 1.5}
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:87]:
```suggestion
                        "condition": {"norm": "lt", "threshold": 1.6}
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 5]
```diff
@@ -50,70 +50,106 @@
...
+                    },
+                    "PCT_OFF_BAIT": {"condition": {"norm": "lt", "threshold": 0.35}},
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:89]:
```suggestion
```
According to the table in amdoc 2046:14, `PCT_OFF_BAIT` criteria is not considered for gicfdna panel

File: BALSAMIC/constants/quality_check_reporting.py [hunk 6]
```diff
@@ -50,70 +50,106 @@
...
+                }
+            },
+            "gmcksolid_4.1_hg19_design.bed": {
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEAN_TARGET_COVERAGE": {
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:94]:
```suggestion
                    "MEDIAN_TARGET_COVERAGE": {
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 7]
```diff
@@ -50,70 +50,106 @@
...
+                        "condition": {"norm": "gt", "threshold": 500}
+                    },
+                    "FOLD_80_BASE_PENALTY": {
+                        "condition": {"norm": "lt", "threshold": 1.7}
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:98]:
```suggestion
                        "condition": {"norm": "lt", "threshold": 1.8}
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 8]
```diff
@@ -50,70 +50,106 @@
...
+                    },
+                    "PCT_OFF_BAIT": {"condition": {"norm": "lt", "threshold": 0.3}},
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:100]:
```suggestion
```
Can you crosscheck again for this criteria

File: BALSAMIC/constants/quality_check_reporting.py [hunk 9]
```diff
@@ -50,70 +50,106 @@
...
+                }
+            },
+            "gmsmyeloid_5.2_hg19_design.bed": {
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEAN_TARGET_COVERAGE": {
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:105]:
```suggestion
                    "MEDIAN_TARGET_COVERAGE": {
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 10]
```diff
@@ -50,70 +50,106 @@
...
+                        "condition": {"norm": "gt", "threshold": 1000}
+                    },
+                    "FOLD_80_BASE_PENALTY": {
+                        "condition": {"norm": "lt", "threshold": 1.5}
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:109]:
```suggestion
                        "condition": {"norm": "lt", "threshold": 1.6}
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 11]
```diff
@@ -50,70 +50,106 @@
...
+                    },
+                    "PCT_OFF_BAIT": {"condition": {"norm": "lt", "threshold": 0.4}},
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:111]:
```suggestion
```

Same as above comment

File: BALSAMIC/constants/quality_check_reporting.py [hunk 12]
```diff
@@ -50,70 +50,106 @@
...
+                }
+            },
+            "lymphoma_6.1_hg19_design.bed": {
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEAN_TARGET_COVERAGE": {
+                        "condition": {"norm": "gt", "threshold": 1000}
+                    },
+                    "FOLD_80_BASE_PENALTY": {
+                        "condition": {"norm": "lt", "threshold": 1.5}
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:120]:
```suggestion
                        "condition": {"norm": "lt", "threshold": 1.6}
```

File: BALSAMIC/constants/quality_check_reporting.py [hunk 13]
```diff
@@ -50,70 +50,106 @@
...
+                    },
+                    "PCT_OFF_BAIT": {"condition": {"norm": "lt", "threshold": 0.35}},
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:122]:
```suggestion
```
Cross-check for this criteria

File: BALSAMIC/constants/quality_check_reporting.py [hunk 14]
```diff
@@ -50,70 +50,106 @@
...
+                }
+            },
+            "gmslymphoid_7.1_hg19_design.bed": {
+                METRIC_FILES["picard_HsMetrics"]: {
+                    "MEAN_TARGET_COVERAGE": {
```
Comment (ashwini06) [inline BALSAMIC/constants/quality_check_reporting.py:127]:
```suggestion
                    "MEDIAN_TARGET_COVERAGE": {
```

File: BALSAMIC/utils/models.py [hunk 15]
```diff
@@ -695,68 +695,60 @@ class BalsamicWorkflowConfig(BaseModel):
...
-        condition: QCConditionModel (optional); condition for QC validation
+
+    Raises:
+        ValidationError: when the specified norm is not supported
+        ValueError: when a metric does not meet its validation requirements
     """
 
     name: str
+    norm: Optional[str] = ...
+    threshold: Optional[float] = ...
```
Comment (ashwini06) [inline BALSAMIC/utils/models.py:714]:
```suggestion
    threshold: Optional[float] = any defaults?
```

File: tests/conftest.py [hunk 16]
```diff
@@ -522,21 +522,37 @@ def qc_extracted_metrics():
...
                     "value": 0.5,
-                    "condition": {"norm": "lt", "threshold": 1.0},
                 },
                 {
                     "name": "MEAN_INSERT_SIZE_2",
+                    "norm": "lt",
+                    "threshold": 1.0,
                     "value": 0.5,
-                    "condition": {"norm": "lt", "threshold": 1.0},
                 },
```
Comment (ashwini06) [inline tests/conftest.py:534]: Are you testing these lines anywhere in the tests? If not, you can maybe remove them?

File: tests/utils/test_qc_metrics.py [hunk 17]
```diff
@@ -72,13 +108,15 @@ def test_get_qc_metrics_dict(analysis_path, qc_metrics):
         "concatenated_tumor": [
             {
                 "name": "MEAN_INSERT_SIZE",
+                "norm": None,
+                "threshold": None,
                 "value": 74.182602,
-                "condition": None,
             },
             {
                 "name": "MEAN_TARGET_COVERAGE",
```
Comment (ashwini06) [inline tests/utils/test_qc_metrics.py:116]:
```suggestion
                "name": "MEDIAN_TARGET_COVERAGE",
```
Analysis QC criteria are primarily based on `median_target_coverage`

**Review by ashwini06 — COMMENTED** [review]
Good job 👍🏼 I have a few comments mostly related to the considered QC criteria. Mainly PCT_OFF_BAIT is no more used for panel qc_check criteria. Also cross-check once again about the chosen thresholds. I might be wrong with my comments, but talk to emma S or Kalle from production to confirm this 😅

File: BALSAMIC/utils/models.py (hunk 15 above)
Comment (ivadym) [reply BALSAMIC/utils/models.py:714]: This is the reason behind this implementation:

> If you want to specify a field that can take a None value while still being required, you can use Optional with ...
> ```
> class Model(BaseModel):
>     a: Optional[int]
>     b: Optional[int] = ...
>     c: Optional[int] = Field(...)
> ```

Source: https://pydantic-docs.helpmanual.io/usage/models/

File: tests/conftest.py (hunk 16 above)
Comment (ivadym) [reply tests/conftest.py:534]: Yes, I am using this fixture to test the models validation.

File: BALSAMIC/constants/quality_check_reporting.py (hunk 1 above)
Comment (ivadym) [reply BALSAMIC/constants/quality_check_reporting.py:76]: It seems that they are being used during the QC report coverage generation (`qc_report.py`) and rendering so for now I'll continue extracting them.

File: BALSAMIC/utils/models.py [hunk 18]
```diff
@@ -695,68 +695,60 @@ class BalsamicWorkflowConfig(BaseModel):
...
 
+    @root_validator(pre=True)
+    def check_operator(cls, values):
+        """Validates if the retrieved norm is a valid operator"""
+        if values["norm"]:
+            assert (
+                values["norm"] in VALID_OPS
+            ), f"{values['norm']} is not a valid condition for QC filtering"
+
+        return values
```
Comment (hassanfa) [inline BALSAMIC/utils/models.py:725]: It's good that you have this, but it might be a bit redundant, no? This looks like a unittest's job.

**Review by hassanfa — COMMENTED** [review]
💯 

Only a small comment on one of the models

File: BALSAMIC/utils/models.py (hunk 18 above)
Comment (ivadym) [reply BALSAMIC/utils/models.py:725]: You're right, if the norm is invalid it will raise an error anyway. Thanks!

**Pull request merged**